VCS fileserver backend fixes/optimizations

terminalmage opened the pull request:
This pull request does the following:
1. Resolves #20785, #17945, #20896, #20993, #21021.
2. Adds the following new fileserver runners: **clear_cache**, **lock**, **clear_lock**, **empty_dir_list**.

In the process of troubleshooting I noticed that dulwich has its own way of handling refs which, when switching back and forth between GitPython/pygit2 and Dulwich can cause additional environments to appear to be available when they are not, particularly (but perhaps not limited to) branches/tags with slashes in their names. It is for this reason that I added the new `fileserver.clear_cache` runner, to make the process of clearing the cache easier.

**EDIT:** In the process of investigating a traceback, I discovered a couple hgfs bugs which affect newer versions of hglib, effectively breaking hgfs. I have fixed these here. In addition, I took the opportunity to add several more fileserver runners.

I was forced to revert the changes from #20141 as they were causing stability issues. I will reopen the original issue (#18839) once this is merged and we will re-investigate.

I also discovered that, while we were laying down a lock file for each repo as we updated it, we were not checking that lockfile before we attempted to fetch those repos again. This could cause unexpected issues if one runs the `fileserver.update` runner during an automated fileserver update. The VCS backends now check for this lockfile before attempting to update a remote.

Finally, I added fixes for minor bugs #17945 and #20993.

Pending test failures and results of code review, this should be ready to merge.

terminalmage committed:
gitfs: fix new branch detection (GitPython)

The existing method of change detection only caught changes to existing
branches. New branches and deleted branches weren't detected as changes.

It turns out that the return value of a fetch with no changes is an
empty list, so simply checking if the return value is non-empty catches
instances where there is a new branch, as well as instances where an
existing branch was updated.

Additionally, the existence of stale refs after a fetch is now used to
determine if a local branch no longer exists on the remote repository.
These stale refs are now deleted to keep the local checkout clean.

terminalmage committed:
gitfs: fix new branch detection (pygit2)

This fixes the same branch/tag detection issues resolved for GitPython
in the previous commit. Additionally, it refines the work done to fix
this issue in GitPython.

terminalmage committed:
Add salt.fileserver.gitfs.clear_cache()

Also rename purge_cache() to more accurately describe what the function
does.

terminalmage committed:
Add salt.fileserver.hgfs.clear_cache()

Also rename purge_cache() to more accurately describe what the function
does.

terminalmage committed:
Add salt.fileserver.svnfs.clear_cache()

Also rename purge_cache() to more accurately describe what the function
does.

terminalmage committed:
Add function in Fileserver class to invoke fsb.clear_cache

terminalmage committed:
Add fileserver.clear_cache runner

terminalmage committed:
Add note about dulwich gitfs cache incompatibility

terminalmage committed:
Add example of clearing gitfs cache pre-2015.2.0

s0undt3ch commented:
Go Go Jenkins!

terminalmage commented:
I am tracking down a traceback that occurs with the write lock stuff that was recently added in a different pull request. See [here](https://github.com/saltstack/salt/issues/20785#issuecomment-75228087) for more information.

Let's hold off on merging until this is resolved.

bechtoldt commented on salt/runners/fileserver.py:
```diff
@@ -89,17 +90,39 @@ def symlink_list(saltenv='base', outputter='nested'):
+
+    .. code-block:: bash
+
+        salt-run fileserver.update
```
update() -> clear_cache()

terminalmage replied:
Good catch, thanks!

terminalmage replied:
Fixed in https://github.com/terminalmage/salt/commit/68815ee, thanks again!

terminalmage committed:
Revert file locking code from PR #20141

terminalmage committed:
Add salt.fileserver.gitfs.clear_lock()

terminalmage committed:
gitfs.clear_cache: return errors instead of ignoring

terminalmage committed:
svnfs.clear_cache: return errors instead of ignoring

terminalmage committed:
hgfs.clear_cache: return errors instead of ignoring

terminalmage committed:
fileserver.clear_cache: return success and errors separately

terminalmage committed:
Add function in Fileserver class to invoke fsb.clear_lock

terminalmage committed:
runners.fileserver.clear_cache: display success and errors separately

terminalmage committed:
add fileserver.clear_lock runner

terminalmage committed:
fix broken salt.utils.is_fcntl_available

terminalmage committed:
Allow "fs" versions of VCS backends to work as arguments to fileserver runner functions

i.e. "gitfs" and "git" both work, "hgfs" and "hg", etc. This should help
reduce confusion.

terminalmage committed:
Add salt.fileserver.hgfs.clear_lock()

terminalmage committed:
Add salt.fileserver.svnfs.clear_lock()

terminalmage committed:
Use shorter version of salt-run command in warning

terminalmage committed:
Add logging noting which remote is being fetched

terminalmage committed:
Fix traceback in salt.fileserver.hgfs.init()

Older versions of hglib did not raise an exception when no refs were
found. This commit catches the exception, fixing a traceback that
prevents new remotes from ever being fetched for the first time.

terminalmage committed:
Move hgfs update lock out of hg checkout

Newer hglib freaks out if files that don't belong in the repo are
present. This commit moves the update lock up a level into
$cachedir/hgfs.

terminalmage committed:
Log an error if unexpected files are found in gitfs/hgfs/svnfs cachedir

This loop is designed to remove the cache directories from old remotes.
shutil.rmtree() will throw an OSError, however, if run on something
other than a directory. This commit catches this exception and logs it.
Seeing this error in the log will be a reminder that some new file has
been added to the cachedir to be used by the FS backend, and that it
needs to be added to the exclusion logic above.

terminalmage committed:
gitfs/hgfs/svnfs: Rewrite _clear_old_remotes()

Rather than using a subtractive approach (doing an os.listdir() and then
removing the stuff we don't want to delete), instead use an additive
approach and assemble a list of things we *do* want to delete. This
makes the function easier to read and maintain.

terminalmage committed:
hgfs: Avoid 2nd init() by returning repos from _clear_old_remotes

terminalmage committed:
svnfs: Avoid 2nd init() by returning repos from _clear_old_remotes

terminalmage committed:
Use new clear_lock() function to clear update lock in update()

terminalmage committed:
Support fileserver backend passed in load for fileserver operations

Also support "-backend" notation to exclude a backend from a given
operation.

terminalmage committed:
Add salt.fileserver.hgfs.lock()

Also rewrite salt.fileserver.hgfs.clear_lock() to accept a repo conf
dict instead of a pattern, so that locking and unlocking in
salt.fileserver.hgfs.update() can use the same code as the fileserver
runner.

terminalmage committed:
Add function in Fileserver class to invoke fsb.lock

terminalmage committed:
Add a backend argument for salt.runners.fileserver.{file,symlink}_list

This allows for backends to be excluded when executing these runners.

terminalmage committed:
Add salt.runners.fileserver.{,empty_}dir_list

This rounds out the runner support for the fileserver runner.

terminalmage committed:
Add salt.runners.fileserver.lock()

terminalmage committed:
Improve salt fileserver documentation

terminalmage committed:
Add salt.fileserver.svnfs.lock()

Also rewrite salt.fileserver.svnfs.clear_lock() to accept a repo conf
dict instead of a pattern, so that locking and unlocking in
salt.fileserver.svnfs.update() can use the same code as the fileserver
runner.

terminalmage committed:
Add salt.fileserver.gitfs.lock()

Also rewrite salt.fileserver.gitfs.clear_lock() to accept a repo conf
dict instead of a pattern, so that locking and unlocking in
salt.fileserver.gitfs.update() can use the same code as the fileserver
runner.

terminalmage committed:
Log success/failure in dealing with lockfiles in their actual functions

Logging them separately in update() doesn't make much sense.

terminalmage committed:
Fix spurious error in master log

When only GitPython is installed, and no gitfs_provider is set, an error
about pygit2 not being available is logged. This should not happen, as
gitfs should just select the first available out of pygit2, gitpython,
and dulwich.

This fixes #17945 by suppressing the error.

terminalmage committed:
Lint fixes

terminalmage committed:
Fix CLI example for fileserver.clear_cache runner

terminalmage committed:
fix missing import

terminalmage committed:
gitfs: Support ssh:// URLs for dulwich

terminalmage committed:
gitfs: Add warning about ssh:// URLs (dulwich)

terminalmage commented:
I just added a fix for #20993 to this PR.

terminalmage commented:
I've got one more change I'd like to get into here: https://github.com/saltstack/salt/issues/21021

terminalmage committed:
Add FileserverConfigError exception class

terminalmage committed:
Fix missing space in log message

terminalmage committed:
gitfs: Raise exceptions on invalid configuration

terminalmage committed:
hg: Raise exceptions on invalid configuration

terminalmage committed:
svnfs: Raise exceptions on invalid configuration

terminalmage committed:
Catch FileserverConfigError exceptions on master startup

This causes the master to fail to start when there are fileserver config
issues, leading people quicker to the log to investigate, where they
will find a meaningful error message. This should it much quicker and
easier for users to realize there are gitfs/hgfs/svnfs fileserver config
problems.

Fixes #21021.

terminalmage commented:
I've pushed a fix for #21021 here as well, waiting on tests.

terminalmage commented:
Aside from the one flaky test, they've all passed. I've re-added the "Tests Passed" tag.
